### PR TITLE
adding modal popup

### PR DESCRIPTION
**Pull request by rasika**

### CODE REVIEW

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag [hunk 1]
```diff
@@ -88,8 +88,26 @@ if (!user) {
         } else {
             //URL: https://localhost:9443/{deviceType}/download?owner={username}
             deviceManagerService = devicemgtProps["httpsURL"] + "/" + deviceType + "/manager";
-
             sketchDownloadEndPoint = deviceManagerService + "/device/" + sketchType + "/download";
+
+            //check if config exists
+            var deviceTypeConfigFile = new File("../app/units/cdmf.unit.device.type."+deviceType+".type-view/private/conf/device-type.json");
+            if(deviceTypeConfigFile.isExists()){
```
Comment (harshanL) [inline features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag:95]: Check the formatting in this if block.

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag [hunk 2]
```diff
@@ -88,8 +88,26 @@ if (!user) {
...
+                deviceTypeConfigFile.open("r");
+                try {
+                    var deviceTypeConfig = parse(deviceTypeConfigFile.readAll());
+                    //replace download endpoint
+                    var downloadAgentUri = deviceTypeConfig.deviceType.downloadAgentUri;
```
Comment (harshanL) [inline features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag:100]: Check the formatting in this if block.

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag [hunk 3]
```diff
@@ -109,8 +127,26 @@ if (!user) {
         } else {
             //URL: https://localhost:9443/{deviceType}/download?owner={username}
             deviceManagerService = devicemgtProps["httpsURL"] + "/" + deviceType + "/manager";
-
             sketchGenerateLinkEndPoint = deviceManagerService + "/device/" + sketchType + "/generate_link";
+
+            //check if config exists
+            var deviceTypeConfigFile = new File("../app/units/cdmf.unit.device.type."+deviceType+".type-view/private/conf/device-type.json");
+            if(deviceTypeConfigFile.isExists()){
```
Comment (harshanL) [inline features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag:134]: Check the formatting in this if block.

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag [hunk 4]
```diff
@@ -88,8 +88,26 @@ if (!user) {
...
+                deviceTypeConfigFile.open("r");
+                try {
+                    var deviceTypeConfig = parse(deviceTypeConfigFile.readAll());
+                    //replace download endpoint
+                    var downloadAgentUri = deviceTypeConfig.deviceType.downloadAgentUri;
+                    if(downloadAgentUri){
+                        sketchDownloadEndPoint = downloadAgentUri;
+                    }
+                }catch(err){
```
Comment (prabathabey) [inline features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag:104]: Fix formatting issues

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag [hunk 5]
```diff
@@ -88,8 +88,26 @@ if (!user) {
...
+                    log.error("Error while reading device config file for `" + deviceType + "`: " + err);
+                }finally{
```
Comment (prabathabey) [inline features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag:106]: More formatting issues.

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag [hunk 6]
```diff
@@ -109,8 +127,26 @@ if (!user) {
...
+            if(deviceTypeConfigFile.isExists()){
+                deviceTypeConfigFile.open("r");
+                try {
+                    var deviceTypeConfig = parse(deviceTypeConfigFile.readAll());
+                    //replace download endpoint
+                    var downloadAgentLinkGenUri = deviceTypeConfig.deviceType.downloadAgentLinkGenUri;
+                    if(downloadAgentUri){
+                        sketchGenerateLinkEndPoint = downloadAgentLinkGenUri;
+                    }
+                }catch(err){
```
Comment (prabathabey) [inline features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag:143]: Fix formatting issues

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag [hunk 7]
```diff
@@ -109,8 +127,26 @@ if (!user) {
...
+                    log.error("Error while reading device config file for `" + deviceType + "`: " + err);
+                }finally{
```
Comment (prabathabey) [inline features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag:145]: Fix formatting issues

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag (hunk 1 above)
Comment (rasika) [inline features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag:95]: this is already fixed with [#6c75515](https://github.com/wso2/carbon-device-mgt/commit/6c75515dc4d31e1c5f63845ee1d2f6a70bc1b156).

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag (hunk 2 above)
Comment (rasika) [inline features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag:100]: this is already fixed with [#6c75515](https://github.com/wso2/carbon-device-mgt/commit/6c75515dc4d31e1c5f63845ee1d2f6a70bc1b156).

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag (hunk 4 above)
Comment (rasika) [inline features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag:104]: this is already fixed with [#6c75515](https://github.com/wso2/carbon-device-mgt/commit/6c75515dc4d31e1c5f63845ee1d2f6a70bc1b156).

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag (hunk 5 above)
Comment (rasika) [inline features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag:106]: this is already fixed with [#6c75515](https://github.com/wso2/carbon-device-mgt/commit/6c75515dc4d31e1c5f63845ee1d2f6a70bc1b156).

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag (hunk 3 above)
Comment (rasika) [inline features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag:134]: this is already fixed with [#6c75515](https://github.com/wso2/carbon-device-mgt/commit/6c75515dc4d31e1c5f63845ee1d2f6a70bc1b156).

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag (hunk 6 above)
Comment (rasika) [inline features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag:143]: this is already fixed with [#6c75515](https://github.com/wso2/carbon-device-mgt/commit/6c75515dc4d31e1c5f63845ee1d2f6a70bc1b156).

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag (hunk 7 above)
Comment (rasika) [inline features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag:145]: this is already fixed with [#6c75515](https://github.com/wso2/carbon-device-mgt/commit/6c75515dc4d31e1c5f63845ee1d2f6a70bc1b156).

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag [hunk 8]
```diff
@@ -109,8 +127,26 @@ if (!user) {
         } else {
             //URL: https://localhost:9443/{deviceType}/download?owner={username}
             deviceManagerService = devicemgtProps["httpsURL"] + "/" + deviceType + "/manager";
-
             sketchGenerateLinkEndPoint = deviceManagerService + "/device/" + sketchType + "/generate_link";
+
+            //check if config exists
+            var deviceTypeConfigFile = new File("../app/units/cdmf.unit.device.type."+deviceType+".type-view/private/conf/device-type.json");
```
Comment (DilanUA) [inline features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag:133]: Fix formatting issues with proper spacing of the string concatenation

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag [hunk 9]
```diff
@@ -109,8 +127,26 @@ if (!user) {
...
+            if(deviceTypeConfigFile.isExists()){
+                deviceTypeConfigFile.open("r");
+                try {
+                    var deviceTypeConfig = parse(deviceTypeConfigFile.readAll());
+                    //replace download endpoint
+                    var downloadAgentLinkGenUri = deviceTypeConfig.deviceType.downloadAgentLinkGenUri;
+                    if(downloadAgentUri){
```
Comment (DilanUA) [inline features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag:140]: Fix formatting issues -> should be like -> if (downloadAgentUri) {

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/app/units/cdmf.unit.ui.theme/public/css/modal.css [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    left: 40px;
+    padding: 5px 8px;
+    font-size: 18px;
+}
+
+.wr-modalpopup .modalpopup-close-btn:hover {
+
+}
+
+/*.wr-modalpopup .modalpopup-content .buttons {*/
```
Comment (Kamidu) [inline features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/app/units/cdmf.unit.ui.theme/public/css/modal.css:65]: Remove commented lines

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag [hunk 11]
```diff
@@ -88,8 +88,26 @@ if (!user) {
         } else {
             //URL: https://localhost:9443/{deviceType}/download?owner={username}
             deviceManagerService = devicemgtProps["httpsURL"] + "/" + deviceType + "/manager";
-
             sketchDownloadEndPoint = deviceManagerService + "/device/" + sketchType + "/download";
+
+            //check if config exists
+            var deviceTypeConfigFile = new File("../app/units/cdmf.unit.device.type."+deviceType+".type-view/private/conf/device-type.json");
```
Comment (DilanUA) [inline features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag:94]: fix formatting issues (no spacing) with this string concatenation.

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag [hunk 12]
```diff
@@ -109,8 +127,26 @@ if (!user) {
         } else {
             //URL: https://localhost:9443/{deviceType}/download?owner={username}
             deviceManagerService = devicemgtProps["httpsURL"] + "/" + deviceType + "/manager";
-
             sketchGenerateLinkEndPoint = deviceManagerService + "/device/" + sketchType + "/generate_link";
+
+            //check if config exists
+            var deviceTypeConfigFile = new File("../app/units/cdmf.unit.device.type."+deviceType+".type-view/private/conf/device-type.json");
+            if(deviceTypeConfigFile.isExists()){
+                deviceTypeConfigFile.open("r");
```
Comment (DilanUA) [inline features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag:135]: shouldn't this line be also inside the try block?

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag (hunk 1 above)
Comment (hasuniea) [inline features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag:95]: Start the first letter with a capital.

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag (hunk 9 above)
Comment (rasika) [inline features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag:140]: This Jaggery API is already refactored in [#6c75515](https://github.com/wso2/carbon-device-mgt/commit/6c75515dc4d31e1c5f63845ee1d2f6a70bc1b156).

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag (hunk 11 above)
Comment (rasika) [inline features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag:94]: This Jaggery API is already refactored in [#6c75515](https://github.com/wso2/carbon-device-mgt/commit/6c75515dc4d31e1c5f63845ee1d2f6a70bc1b156).

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag (hunk 8 above)
Comment (rasika) [inline features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag:133]: This Jaggery API is already refactored in [#6c75515](https://github.com/wso2/carbon-device-mgt/commit/6c75515dc4d31e1c5f63845ee1d2f6a70bc1b156).

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag (hunk 1 above)
Comment (rasika) [inline features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag:95]: This comment is no longer exists

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/app/units/cdmf.unit.ui.theme/public/css/modal.css (hunk 10 above)
Comment (rasika) [inline features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/app/units/cdmf.unit.ui.theme/public/css/modal.css:65]: Fixed with [#0ebede6](https://github.com/wso2/carbon-device-mgt/commit/0ebede64989d0c09dab5c3cb9b8b955d53fc98b6).

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag (hunk 12 above)
Comment (rasika) [inline features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/api/device-api.jag:135]: Agreed. This logic has already moved to utility.js and fixed with [#01342e0](https://github.com/wso2/carbon-device-mgt/commit/01342e03f5516ca9a93ec077fe3e457a25d65137)

**Pull request merged**